File: frontend/src/data/scoring.ts
// Formula port of bot/scoring.py — must stay in sync. Buckets, weights, and pod
// point values come from the shared scoring_buckets.json (same file Python loads).
//
//   raw_group  = weighted_trophies × weight × trophy_rate
//   confidence = T / (T + 2)        // T = total trophies across all groups
//   total      = (Σ raw_group) × confidence
//
// Confidence is aggregate (one factor over total trophies), and LCQ Draft 2 keeps
// its wins×winrate×weight rule, exempt from confidence. Pod points are a separate
// flat term added by callers via podPoints().
//
// weightedTrophies counts a trophy by what the end-of-event rank was worth, and the
// database supplies it. Only the numerator is weighted: trophy rate and the confidence
// total stay on raw counts. It falls back to trophies where a caller has no rank data.
import {
  BUCKET_DEFS,
  formatsForBucket,
  POD_TROPHY_POINTS,
  POD_WIN_2_1_POINTS,
} from "./format-buckets";

interface QueueGroup {
  label: string;
  points: number;
  formats: readonly string[];
  rule?: "lcq_draft_2";
}

const DEFAULT_QUEUE_GROUPS: readonly QueueGroup[] = BUCKET_DEFS.map((d) => ({
  ...d,
  formats: formatsForBucket(d.label),
}));

export interface ScoringStatRow {
  format: string;
  wins: number;
  losses: number;
  trophies: number;
  weightedTrophies?: number;
  events: number;
}

export interface GroupTotals {
  label: string;
  events: number;
  wins: number;
  losses: number;
  trophies: number;
  weightedTrophies?: number;
}

export interface Aggregate {
  total: number;
  confidence: number;
  contributionByLabel: Map<string, number>;
}

export function podPoints(trophies: number, wins21: number): number {
  return trophies * POD_TROPHY_POINTS + wins21 * POD_WIN_2_1_POINTS;
}

export function confidenceFactor(totalTrophies: number): number {
  return totalTrophies > 0 ? totalTrophies / (totalTrophies + 2) : 0;
}

function defFor(label: string): QueueGroup | undefined {
  return DEFAULT_QUEUE_GROUPS.find((g) => g.label === label);
}

// Per-group contribution (already × aggregate confidence) + the confidence factor.
// Input is per-group totals; rows for the same label are summed first.
export function aggregate(groups: GroupTotals[]): Aggregate {
  const byLabel = new Map<string, GroupTotals>();
  for (const g of groups) {
    if (!defFor(g.label)) continue;
    const cur = byLabel.get(g.label);
    if (cur) {
      cur.events += g.events;
      cur.wins += g.wins;
      cur.losses += g.losses;
      cur.trophies += g.trophies;
      cur.weightedTrophies = (cur.weightedTrophies ?? 0) + (g.weightedTrophies || g.trophies);
    } else {
      byLabel.set(g.label, { ...g, weightedTrophies: g.weightedTrophies || g.trophies });
    }
  }

  const rawByLabel = new Map<string, number>();
  const lcqByLabel = new Map<string, number>();
  let totalTrophies = 0;
  for (const def of DEFAULT_QUEUE_GROUPS) {
    const g = byLabel.get(def.label);
    if (!g) continue;
    if (def.rule === "lcq_draft_2") {
      const games = g.wins + g.losses;
      if (games > 0 && g.wins > 0) {
        lcqByLabel.set(def.label, g.wins * (g.wins / games) * def.points);
      }
      continue;
    }
    if (g.trophies === 0 || g.events === 0) continue;
    const weighted = g.weightedTrophies || g.trophies;
    rawByLabel.set(def.label, weighted * def.points * (g.trophies / g.events));
    totalTrophies += g.trophies;
  }

  const confidence = confidenceFactor(totalTrophies);
  const contributionByLabel = new Map<string, number>();
  for (const [label, raw] of rawByLabel) contributionByLabel.set(label, raw * confidence);
  for (const [label, score] of lcqByLabel) contributionByLabel.set(label, score);

  let total = 0;
  for (const v of contributionByLabel.values()) total += v;
  return { total, confidence, contributionByLabel };
}

export function groupTotalsFromRows(rows: ScoringStatRow[]): GroupTotals[] {
  const byLabel = new Map<string, GroupTotals>();
  for (const row of rows) {
    const def = DEFAULT_QUEUE_GROUPS.find((g) => g.formats.includes(row.format));
    if (!def) continue;
    const cur =
      byLabel.get(def.label) ??
      { label: def.label, events: 0, wins: 0, losses: 0, trophies: 0, weightedTrophies: 0 };
    cur.events += row.events;
    cur.wins += row.wins;
    cur.losses += row.losses;
    cur.trophies += row.trophies;
    cur.weightedTrophies = (cur.weightedTrophies ?? 0) + (row.weightedTrophies || row.trophies);
    byLabel.set(def.label, cur);
  }
  return [...byLabel.values()];
}

// A public_*_breakdown row already carries its group label, so it becomes GroupTotals directly.
export function groupTotalsFromBreakdown(
  rows: readonly {
    formatLabel: string;
    events: number;
    wins: number;
    losses: number;
    trophies: number;
    weightedTrophies?: number;
  }[],
): GroupTotals[] {
  return rows.map((r) => ({
    label: r.formatLabel,
    events: r.events,
    wins: r.wins,
    losses: r.losses,
    trophies: r.trophies,
    weightedTrophies: r.weightedTrophies,
  }));
}

export function computeScore(rows: ScoringStatRow[]): number {
  return Math.round(aggregate(groupTotalsFromRows(rows)).total * 100) / 100;
}

export function scoreFromGroups(groups: GroupTotals[]): number {
  return Math.round(aggregate(groups).total * 100) / 100;
}

// Arena Direct box payouts — port of bot/scoring.py boxes_for_event and the era
// constants in bot/sets.py. Must stay in sync.
const SIX_WIN_PLAY_DIRECT_SETS = new Set(["OTJ", "FDN", "BLB", "DSK"]);
const SIX_WIN_COLLECTOR_DIRECT_SETS = new Set(["DFT"]);

const COLLECTOR_BOOSTER_WINDOWS: ReadonlyArray<{ setCode: string; startDate: string; endDate: string }> = [
  { setCode: "TDM", startDate: "2025-04-18", endDate: "2025-04-20" },
  { setCode: "FIN", startDate: "2025-06-20", endDate: "2025-06-22" },
  { setCode: "EOE", startDate: "2025-08-08", endDate: "2025-08-11" },
  { setCode: "TLA", startDate: "2025-11-28", endDate: "2025-11-30" },
  { setCode: "ECL", startDate: "2026-01-30", endDate: "2026-02-01" },
  { setCode: "TMT", startDate: "2026-03-13", endDate: "2026-03-15" },
  { setCode: "SOS", startDate: "2026-04-30", endDate: "2026-05-04" },
  { setCode: "MSH", startDate: "2026-06-30", endDate: "2026-07-06" },
  { setCode: "HOB", startDate: "2026-08-14", endDate: "2026-08-23" },
];

const COLLECTOR_WINDOW_SLACK_DAYS = 1;

function shiftDays(isoDate: string, days: number): string {
  const d = new Date(`${isoDate}T00:00:00Z`);
  d.setUTCDate(d.getUTCDate() + days);
  return d.toISOString().slice(0, 10);
}

function isCollectorBoosterWindow(setCode: string, finishedAt: string): boolean {
  const day = finishedAt.slice(0, 10);
  for (const w of COLLECTOR_BOOSTER_WINDOWS) {
    if (w.setCode !== setCode) continue;
    const start = shiftDays(w.startDate, -COLLECTOR_WINDOW_SLACK_DAYS);
    const end = shiftDays(w.endDate, COLLECTOR_WINDOW_SLACK_DAYS);
    if (start <= day && day <= end) return true;
  }
  return false;
}

export function boxesForEvent(
  setCode: string,
  wins: number,
  finishedAt: string | null,
  isTrophy: boolean,
): number {
  if (SIX_WIN_PLAY_DIRECT_SETS.has(setCode)) return isTrophy ? 2 : 0;
  if (SIX_WIN_COLLECTOR_DIRECT_SETS.has(setCode)) return isTrophy ? 1 : 0;
  if (finishedAt && isCollectorBoosterWindow(setCode, finishedAt)) return isTrophy ? 1 : 0;
  return wins >= 7 ? 2 : wins === 6 ? 1 : 0;
}

// LCQ Draft 2 cash payouts per event: a 6-win run pays $2k, a 5-win run $1k.
export function lcqDraft2Earnings(wins: number): number {
  return wins >= 6 ? 2000 : wins === 5 ? 1000 : 0;
}

File: bot/scoring.py
"""Format groups and scoring formula for player rating.

    raw_group   = weighted_trophies × group_points × trophy_rate
    confidence  = T / (T + 2)        # T = total trophies across all groups
    total       = (Σ raw_group) × confidence

with a special case for LCQ Draft 2 (counts wins instead of trophies, uses
winrate, and is exempt from the confidence multiplier).

Confidence is an aggregate: it shrinks the whole resume by one factor built
from total trophy count, rather than penalising each queue group on its own
small sample. Pod-draft points are a separate flat term (see ``pod_points``),
added to the leaderboard total outside this module.

``weighted_trophies`` counts a trophy by what the player's rank was worth at
the end of the event, so a group carrying ``rank_points`` pays more for a
trophy taken at a high rank. Only the numerator is weighted: trophy_rate and
the confidence total stay on raw counts, so a rank weight moves the multiplier
and never the rate. Groups without ``rank_points`` weight every trophy at 1.0,
which leaves ``weighted_trophies`` equal to ``trophies``. A row carrying trophies
but no weighted count scores unweighted, so a writer that predates the column
degrades to the old formula instead of to zero.

A ``QueueGroup`` collects raw 17lands ``format`` strings under a single label
that shares a points weight. The ``Sealed`` group rolls best-of-1 and
best-of-3 sealed under one entry — sealed is sealed for scoring purposes. LCQ
groups sit dormant until WOTC actually runs LCQ events for the season.
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from datetime import datetime
from pathlib import Path
from typing import Iterable, Mapping, Sequence

from bot.sets import (
    SIX_WIN_COLLECTOR_DIRECT_SETS,
    SIX_WIN_PLAY_DIRECT_SETS,
    is_collector_booster_window,
)


@dataclass(frozen=True)
class QueueGroup:
    label: str
    points: int
    formats: tuple[str, ...]
    # None = standard formula; "lcq_draft_2" = wins-not-trophies × winrate × points
    rule: str | None = None
    # Arena rank tier -> what a trophy taken at that tier is worth, in the same points unit
    rank_points: Mapping[str, int] = field(default_factory=dict)

    def trophy_weight(self, end_rank: str | None) -> float:
        """Multiplier for one trophy, keyed on the tier of a ``Platinum-3``-shaped rank string."""
        if not end_rank or not self.rank_points:
            return 1.0
        tier = end_rank.split("-")[0]
        return self.rank_points.get(tier, self.points) / self.points


BUCKETS_JSON = Path(__file__).resolve().parents[1] / "scoring_buckets.json"
_BUCKETS = json.loads(BUCKETS_JSON.read_text())

DEFAULT_QUEUE_GROUPS: tuple[QueueGroup, ...] = tuple(
    QueueGroup(
        label=g["label"],
        points=int(g["points"]),
        formats=tuple(g["formats"]),
        rule=g.get("rule"),
        rank_points={tier: int(pts) for tier, pts in g.get("rank_points", {}).items()},
    )
    for g in _BUCKETS["groups"]
)

POD_TROPHY_POINTS = float(_BUCKETS.get("pod", {}).get("trophy_points", 5))
POD_TWO_WIN_POINTS = float(_BUCKETS.get("pod", {}).get("two_win_points", 2))
POD_ONE_WIN_POINTS = float(_BUCKETS.get("pod", {}).get("one_win_points", 0.5))

ARENA_DIRECT_SEALED_FORMAT = "ArenaDirect_Sealed"


def pod_points(trophies: int, two_win_finishes: int, one_win_finishes: int) -> int:
    """Flat pod-draft contribution, floored so fractional finishes bank across pods.

    Paid on match wins, not on the record string: a trophy is a 3-0 or a pod win, and two wins pay the
    same whether the third match was lost or forfeited by a drop. Added to the leaderboard total
    alongside ``compute_score`` — pods are not a 17lands queue group and are exempt from
    trophy_rate / confidence.
    """
    total = (
        trophies * POD_TROPHY_POINTS
        + two_win_finishes * POD_TWO_WIN_POINTS
        + one_win_finishes * POD_ONE_WIN_POINTS
    )
    return int(total)


def supported_formats(groups: Iterable[QueueGroup] = DEFAULT_QUEUE_GROUPS) -> tuple[str, ...]:
    return tuple(fmt for g in groups for fmt in g.formats)


def group_label_for_format(fmt: str | None, groups: Iterable[QueueGroup] = DEFAULT_QUEUE_GROUPS) -> str | None:
    if fmt is None:
        return None
    for group in groups:
        if fmt in group.formats:
            return group.label
    return None


def confidence_factor(total_trophies: int) -> float:
    """Aggregate shrinkage prior: T/(T+2) over the player's whole trophy count."""
    return total_trophies / (total_trophies + 2) if total_trophies > 0 else 0.0


def _aggregate(
    stats_rows: Sequence[dict],
    groups: Iterable[QueueGroup],
) -> tuple[dict[str, float], float]:
    """Per-group score contribution (by label) and the aggregate confidence factor.

    Non-LCQ groups contribute ``raw_group × confidence`` where confidence is built
    from total trophies across all non-LCQ groups, so the parts sum to the total.
    LCQ Draft 2 keeps its wins×winrate×points rule and is exempt from confidence.
    """
    groups = list(groups)
    grouped: dict[str, list[dict]] = {}
    for row in stats_rows:
        g = _group_for_format(groups, row["format"])
        if g is not None:
            grouped.setdefault(g.label, []).append(row)

    raw_by_label: dict[str, float] = {}
    lcq_by_label: dict[str, float] = {}
    total_trophies = 0
    for g in groups:
        rows = grouped.get(g.label)
        if not rows:
            continue
        if g.rule == "lcq_draft_2":
            wins = sum(r.get("wins", 0) for r in rows)
            losses = sum(r.get("losses", 0) for r in rows)
            games = wins + losses
            if games and wins:
                lcq_by_label[g.label] = wins * (wins / games) * g.points
            continue
        trophies = sum(r.get("trophies", 0) for r in rows)
        events = sum(r.get("events", 0) for r in rows)
        if trophies == 0 or events == 0:
            continue
        weighted = sum(r.get("weighted_trophies") or r.get("trophies", 0) for r in rows)
        raw_by_label[g.label] = weighted * g.points * (trophies / events)
        total_trophies += trophies

    confidence = confidence_factor(total_trophies)
    contrib = {label: raw * confidence for label, raw in raw_by_label.items()}
    contrib.update(lcq_by_label)
    return contrib, confidence


def compute_score_breakdown(
    stats_rows: Sequence[dict],
    groups: Iterable[QueueGroup] = DEFAULT_QUEUE_GROUPS,
) -> list[dict]:
    """Per-group totals + score contribution, derived from the aggregate so the
    contributions sum to ``compute_score``. Skips groups with no matching rows.
    """
    groups_list = list(groups)
    grouped: dict[str, list[dict]] = {}
    for row in stats_rows:
        g = _group_for_format(groups_list, row["format"])
        if g is not None:
            grouped.setdefault(g.label, []).append(row)

    contrib, _ = _aggregate(stats_rows, groups_list)
    breakdown: list[dict] = []
    for g in groups_list:
        rows = grouped.get(g.label)
        if not rows:
            continue
        breakdown.append({
            "label": g.label,
            "events": sum(r.get("events", 0) for r in rows),
            "wins": sum(r.get("wins", 0) for r in rows),
            "losses": sum(r.get("losses", 0) for r in rows),
            "trophies": sum(r.get("trophies", 0) for r in rows),
            "score": round(contrib.get(g.label, 0.0), 2),
        })
    return breakdown


def compute_score(
    stats_rows: Sequence[dict],
    groups: Iterable[QueueGroup] = DEFAULT_QUEUE_GROUPS,
) -> float:
    """Total 17lands score for one player across all their group-rolled stats.

    ``stats_rows`` items are dicts with keys: format, wins, losses, trophies, events,
    and optionally weighted_trophies, which falls back to trophies when absent.
    Rows whose format isn't in any group are ignored. Pod points are added by the
    caller via ``pod_points``.
    """
    contrib, _ = _aggregate(stats_rows, groups)
    return round(sum(contrib.values()), 2)


def boxes_for_event(set_code: str, wins: int, finished_at: datetime | None, is_trophy: bool) -> int:
    """Boxes awarded for a single Arena Direct Sealed event, per the era rules in bot.sets.

    A collector premiere always pays one box for winning the event, and a 2024 play set
    pays two — both keyed on ``is_trophy`` so the per-era win cap (6 vs 7, including the
    mid-TDM rollout) comes straight from 17lands rather than a hardcoded threshold. Only
    the default 7-win play ladder needs raw wins, for its 1-box consolation at six.
    """
    if set_code in SIX_WIN_PLAY_DIRECT_SETS:
        return 2 if is_trophy else 0
    if set_code in SIX_WIN_COLLECTOR_DIRECT_SETS:
        return 1 if is_trophy else 0
    if finished_at is not None and is_collector_booster_window(set_code, finished_at.date()):
        return 1 if is_trophy else 0
    return 2 if wins >= 7 else (1 if wins == 6 else 0)


def _group_for_format(groups: Iterable[QueueGroup], fmt: str) -> QueueGroup | None:
    for g in groups:
        if fmt in g.formats:
            return g
    return None
